File: src/main/java/fr/spoutnik87/musicbot_rest/security/JWTAuthenticationFilter.java
package fr.spoutnik87.musicbot_rest.security;

import com.auth0.jwt.JWT;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import fr.spoutnik87.musicbot_rest.model.User;
import fr.spoutnik87.musicbot_rest.model.Views;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Date;

import static com.auth0.jwt.algorithms.Algorithm.HMAC512;

public class JWTAuthenticationFilter extends UsernamePasswordAuthenticationFilter {

  private UserDetailsServiceImpl userDetailsService;
  private AuthenticationManager authenticationManager;

  private SecurityConfiguration securityConfiguration;

  public JWTAuthenticationFilter(
      UserDetailsServiceImpl userDetailsService,
      AuthenticationManager authenticationManager,
      SecurityConfiguration securityConfiguration) {
    this.userDetailsService = userDetailsService;
    this.authenticationManager = authenticationManager;
    this.securityConfiguration = securityConfiguration;
  }

  @Override
  public Authentication attemptAuthentication(HttpServletRequest req, HttpServletResponse res)
      throws AuthenticationException {
    try {
      fr.spoutnik87.musicbot_rest.security.User creds =
          new ObjectMapper()
              .readValue(req.getInputStream(), fr.spoutnik87.musicbot_rest.security.User.class);

      return authenticationManager.authenticate(
          new UsernamePasswordAuthenticationToken(
              creds.getEmail(),
              creds.getPassword(),
              userDetailsService.loadUserByUsername(creds.getEmail()).getAuthorities()));
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  protected void successfulAuthentication(
      HttpServletRequest req, HttpServletResponse res, FilterChain chain, Authentication auth) {
    String token =
        JWT.create()
            .withSubject(((UserDetails) auth.getPrincipal()).getUsername())
            .withExpiresAt(
                new Date(
                    System.currentTimeMillis() + this.securityConfiguration.getExpirationTime()))
            .sign(HMAC512(this.securityConfiguration.getSecret().getBytes()));
    res.addHeader("Access-Control-Expose-Headers", "Authorization");
    res.addHeader(
        this.securityConfiguration.getHeaderString(),
        this.securityConfiguration.getTokenPrefix() + token);
    res.setContentType("application/json");
    res.setCharacterEncoding("UTF-8");
    try {
      ObjectMapper mapper = new ObjectMapper();
      User u = ((UserDetails) auth.getPrincipal()).getUser();
      mapper.disable(MapperFeature.DEFAULT_VIEW_INCLUSION);
      String result = mapper.writerWithView(Views.Public.class).writeValueAsString(u);
      res.getWriter().print(result);
      res.getWriter().flush();
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}

File: src/main/kotlin/fr/spoutnik87/musicbot_rest/security/JWTAuthenticationFilter.kt
package fr.spoutnik87.musicbot_rest.security

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm.HMAC512
import com.fasterxml.jackson.databind.MapperFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import fr.spoutnik87.musicbot_rest.model.Views
import fr.spoutnik87.musicbot_rest.viewmodel.UserViewModel
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import java.io.IOException
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JWTAuthenticationFilter(
        private val userDetailsService: UserDetailsServiceImpl,
        private val authManager: AuthenticationManager,
        private val securityConfiguration: SecurityConfiguration
) : UsernamePasswordAuthenticationFilter() {

    @Throws(RuntimeException::class)
    override fun attemptAuthentication(req: HttpServletRequest?, res: HttpServletResponse?): Authentication {
        try {
            val creds = ObjectMapper().readValue(req?.inputStream, User::class.java)
            return authManager.authenticate(
                    UsernamePasswordAuthenticationToken(creds.email, creds.password,
                            userDetailsService.loadUserByUsername(creds.email)?.authorities))
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun successfulAuthentication(req: HttpServletRequest, res: HttpServletResponse, chain: FilterChain?, auth: Authentication?) {
        val username = (auth?.principal as org.springframework.security.core.userdetails.UserDetails).username
        val token = JWT.create()
                .withSubject(username)
                .withExpiresAt(Date(System.currentTimeMillis() + this.securityConfiguration.expirationTime))
                .sign(HMAC512(this.securityConfiguration.secret.toByteArray()))
        res.addHeader("Access-Control-Expose-Headers", "Authorization")
        res.addHeader(
                this.securityConfiguration.headerString,
                this.securityConfiguration.tokenPrefix + token)
        res.contentType = "application/json"
        res.characterEncoding = "UTF-8"

        val mapper = ObjectMapper().registerModule(KotlinModule())
        mapper.disable(MapperFeature.DEFAULT_VIEW_INCLUSION)
        val user = userDetailsService.loadUserByUsername(username)?.user
        res.writer.print(mapper.writerWithView(Views.Companion.Mixed::class.java)
                .writeValueAsString(UserViewModel.from(user!!)))
        res.writer.flush()
    }
}